exla/lib/exla/defn.ex

```diff
@@ -2686,28 +2710,64 @@ defmodule EXLA.Defn do
          cache,
          fun
        ) do
-    # TO-DO(mlir): deal with token
-    inputs = Enum.with_index(args, fn arg, idx -> {"p#{idx}", Value.get_shape(arg)} end)
+    if token = get_token(cache) do
+      # TO-DO(mlir): deal with token
+      inputs = Enum.with_index(args, fn arg, idx -> {"p#{idx + 1}", Value.get_shape(arg)} end)
+      inputs = [{"p0", EXLA.Shape.make_token_shape()} | inputs]
+
+      # input function is actually the parent function still, so we need to actually create a new function
+      # with this name on the same module.
+      function =
+        EXLA.Builder.new(
+          {module, name},
+          inputs,
+          {struct(Nx.Tensor, type: :token), out_expr},
+          :mlir,
+          false,
+          true
+        )
 
-    # input function is actually the parent function still, so we need to actually create a new function
-    # with this name on the same module.
-    function = EXLA.Builder.new({module, name}, inputs, out_expr, :mlir, false, true)
+      case function.return_shape do
+        [%{dtype: {:tuple, _}}] ->
+          raise "MLIR cannot return tuple from if branch"
```

Comment:
In the future, we should not need this check, because we always flatten it before calling the builder.

Reply:
Yes :)